core/block/undo/undo.go

```diff
@@ -124,10 +133,16 @@ func (h *history) Len() int {
 	return h.pointer
 }
 
-func (h *history) Previous() (Action, error) {
+func (h *history) Previous() (action Action, err error) {
 	if h.pointer > 0 {
 		h.pointer--
-		return h.actions[h.pointer], nil
+		action = h.actions[h.pointer]
+		if h.pointer > 0 {
+			action.CarriageInfo = h.actions[h.pointer-1].CarriageInfo
```

Comment:
I'm confused that you update carriage info here. Maybe it's better to return carriage info explicitly? 

Or even better replicate for carriage how Before-After information is stored in Action like Details, RelationLinks?